### ISSUE TITLE
Can't change default parser for Postgresql using setTypeParser after version 4.32.3

### ISSUE THREAD

**Issue by olmo** [sequelize/sequelize#9074]
## What are you doing?

I'm trying to change the default parser for decimal type (numeric in postgresql) from string to float using the following code:
```js
var pg = require('pg');
pg.types.setTypeParser(1700, 'text', parseFloat);
```
This was working until version 4.32.4

## What do you expect to happen?
The same behavior prior to version 4.32.4

## What is actually happening?
The parser doesn't change.

__Dialect:__ postgres
__Dialect version:__ 7.4.1
__Database version:__ 9.6
__Sequelize version:__ 4.33.4

**Note :** _Your issue may be ignored OR closed by maintainers if it's not tested against latest version OR does not follow issue template._

**Comment by Vovka19** [comment]
Thank you!

**Comment by tfazekas** [comment]
Hi olmo,

Same issue here.
Just upgraded to 4.32.4

Did you find any solution?

**Comment by kingjerod** [comment]
After digging through the code, I was able to get it to work with this
```
sequelize.Sequelize.DataTypes.postgres.DECIMAL.parse = parseFloat;
```

Wonder how long this hack will last...

**Comment by psotres** [comment]
I've also been digging though the code and problem seems to have been introduced with #9025 and #9026.

The way the parser is stored for future usage in the connectionManager (via a function that will call ``dataType.parse`` method when needed instead of storing its current reference) makes adding custom parser logic quite obscure if you don't clone the ``dataType`` object, as a subsequent modification of ``Sequelize.DataTypes.postgres.*.parse`` value in another sequelize instance will impact all instances.

My current workaround to solve the problem without having to use the hack provided by @kingjerod (which is a valid solution but it impacts all sequelize instances) is this:
```javascript
const types = {
  DECIMAL: { ...sequelize.Sequelize.DataTypes.postgres.DECIMAL },
};
types.DECIMAL.parse = parseFloat;
sequelize.connectionManager.refreshTypeParser(types);
```

Nevertheless, as a ``getTypeParser(oid)`` method is already there, I would say the more elegant way to provide the functionality is via a new ``setTypeParser(oid, parser)`` method.

**Comment by EyePulp** [comment]
@psotres Your fix did the trick for me.  But wow, what a frustrating red-herring with all the articles pointing to the `setTypeParser(oid, parser)` methods, but having DECIMAL (or NUMERIC in postgres' case) not work the same or be overwritten during instantiation.

Thanks again for figuring this out.

**Comment by k98kumar** [comment]
@EyePulp 
I'm using v5.x, but when I use the hooks, I get:
`Error: Unrecognized datatype for attribute "pgeRating.hooks"` with
```
const rating = sequelize.define('rating', {
  // properties
}, {
  hooks: {
    afterConnect: // Rest of your code
  }
})
```
Instead I used the addHook function:
```
pgeRating.addHook('afterConnect', function setParsers() {
  const cloneType = ...
  ...
  this.connectionManager.refreshTypeParser(dTypes);
}
```
Assuming I used the addHook function correctly, I still got DECIMAL numbers in string format. Am I forgetting to do something or did I not set addHook up correctly?

**Comment by guillaumeduliscouet** [comment]
For anyone facing the same problem as @k98kumar, refer to https://sequelize.org/master/manual/hooks.html. It seems like @EyePulp is referring to  **Permanent Hooks**, they can be added this way for example:
```
new Sequelize(..., {
  hooks: {
    afterConnect() {
      // do stuff
    }
  }
});
```
Anyway, @EyePulp's workaround works for me, thanks

**Comment by Leithal3** [comment]
@EyePulp @psotres 

I would like to start by thanking you, your help was invaluable.

Having a look it seems we can get the workaround to look a little cleaner by extending the class instead of trying to clone the entire object.

Defining our CustomDecimal as follows alllows us to override the parse function and keep all of the other information associated. 

```ts
class CustomDecimal extends (Sequelize.DataTypes.DECIMAL as Sequelize.DataTypes.NumberDataTypeConstructor) {
  static parse (value: string): number {
    return parseFloat(value)
  }
}
```
This enables us to make the hook itself a bit cleaner.
```ts
hooks: {
    afterConnect (): Promise<void> | void {
      const dTypes = {
        DECIMAL: CustomDecimal
      };
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      (this as any as Sequelize.Sequelize).connectionManager.refreshTypeParser(dTypes)
    }
  }
```
May be important to mention that I am using TypeScript. 
Hopefully this helps!

**Comment by EyePulp** [comment]
@Leithal3 Thanks for the more succinct implementation!  I'll admit I'm not entirely sure how this would look as standard ES6, as my familiarity with TypeScript is pretty limited =)

If you're feeling generous, I'd love a translated version, but regardless I appreciate your effort - thanks again.

**Comment by Leithal3** [comment]
@EyePulp - I'm happy to give it a go - but have no way to verify right now.

Removing Typescript related code gives the following:
```js
class CustomDecimal extends Sequelize.DataTypes.DECIMAL {
  static parse (value) {
    return parseFloat(value)
  }
}

hooks: {
  afterConnect: function () {
    const dTypes = {
      DECIMAL: CustomDecimal
    };
    this.connectionManager.refreshTypeParser(dTypes)
  }
}
```

Good luck!

**Comment by EyePulp** [comment]
@Leithal3 That works, thanks so much.  That's pretty much identical to where I ended up.  Clearly the best solution is for us to start using TS more often. =)

**Comment by jobelenus** [comment]
I am on node12 and seq6.3.5. I have all my INTEGERs (even FKs) and BIGINT fields displaying as strings in JSON/get({plain}).

I tried @Leithal3 approach and confirmed that I was parsing ints with my custom parser type. However, the values were still strings in my JSON output.

In order to get them to be integers in a JSON representation, I had to add getters and return the parseInt value from there.

In short, this hook approach did not work for me.

**Comment by ephys** [comment]
The workaround mentioned in this thread are going to break again in v7, but we have worked on a stable solution for this:

Starting with the next v7 alpha release, Data Types are a lot more powerful. Take a look at our [updated documentation](https://sequelize.org/docs/v7/other-topics/extending-data-types/#modifying-an-existing-data-type) to see how you can customize an existing Data Type. [This feature request](https://github.com/sequelize/sequelize/issues/14296) should also cover this use case without needing to use internal APIs.

**Comment by jvalerodev** [comment]
> @EyePulp @psotres
> 
> Me gustaría comenzar agradeciéndole, su ayuda fue invaluable.
> 
> Echando un vistazo, parece que podemos hacer que la solución se vea un poco más limpia extendiendo la clase en lugar de intentar clonar todo el objeto.
> 
> Definir nuestro CustomDecimal de la siguiente manera nos permite anular la función de análisis y mantener toda la otra información asociada.
> 
> ```ts
> class CustomDecimal extends (Sequelize.DataTypes.DECIMAL as Sequelize.DataTypes.NumberDataTypeConstructor) {
>   static parse (value: string): number {
>     return parseFloat(value)
>   }
> }
> ```
> 
> Esto nos permite hacer que el anzuelo sea un poco más limpio.
> 
> ```ts
> hooks: {
>     afterConnect (): Promise<void> | void {
>       const dTypes = {
>         DECIMAL: CustomDecimal
>       };
>       // eslint-disable-next-line @typescript-eslint/no-explicit-any
>       (this as any as Sequelize.Sequelize).connectionManager.refreshTypeParser(dTypes)
>     }
>   }
> ```
> 
> Puede ser importante mencionar que estoy usando TypeScript. ¡Ojalá esto ayude!

I get the following error: "Base constructors must all have the same return type."

Could you tell me how can I fix it?

**Comment by jvalerodev** [comment]
> @EyePulp @psotres
> 
> I would like to start by thanking you, your help was invaluable.
> 
> Having a look it seems we can get the workaround to look a little cleaner by extending the class instead of trying to clone the entire object.
> 
> Defining our CustomDecimal as follows alllows us to override the parse function and keep all of the other information associated.
> 
> ```ts
> class CustomDecimal extends (Sequelize.DataTypes.DECIMAL as Sequelize.DataTypes.NumberDataTypeConstructor) {
>   static parse (value: string): number {
>     return parseFloat(value)
>   }
> }
> ```
> 
> This enables us to make the hook itself a bit cleaner.
> 
> ```ts
> hooks: {
>     afterConnect (): Promise<void> | void {
>       const dTypes = {
>         DECIMAL: CustomDecimal
>       };
>       // eslint-disable-next-line @typescript-eslint/no-explicit-any
>       (this as any as Sequelize.Sequelize).connectionManager.refreshTypeParser(dTypes)
>     }
>   }
> ```
> 
> May be important to mention that I am using TypeScript. Hopefully this helps!

I get the following error: "Base constructors must all have the same return type."

Could you tell me how can I fix it?

**Comment by Chewbye** [comment]
> @EyePulp - I'm happy to give it a go - but have no way to verify right now.
> 
> Removing Typescript related code gives the following:
> 
> ```js
> class CustomDecimal extends Sequelize.DataTypes.DECIMAL {
>   static parse (value) {
>     return parseFloat(value)
>   }
> }
> 
> hooks: {
>   afterConnect: function () {
>     const dTypes = {
>       DECIMAL: CustomDecimal
>     };
>     this.connectionManager.refreshTypeParser(dTypes)
>   }
> }
> ```
> 
> Good luck!

This workaround worked, thank you

**Comment by nikolaydyankov** [comment]
> > @EyePulp - I'm happy to give it a go - but have no way to verify right now.
> > Removing Typescript related code gives the following:
> > ```js
> > class CustomDecimal extends Sequelize.DataTypes.DECIMAL {
> >   static parse (value) {
> >     return parseFloat(value)
> >   }
> > }
> > 
> > hooks: {
> >   afterConnect: function () {
> >     const dTypes = {
> >       DECIMAL: CustomDecimal
> >     };
> >     this.connectionManager.refreshTypeParser(dTypes)
> >   }
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Good luck!
> 
> This workaround worked with version 6.33, thank you

Thank you for the fix, it works. The only thing is that TS complains:

```
Property 'DataTypes' does not exist on type 'typeof Sequelize'.ts(2339)
```

**Comment by seanjdunn85** [comment]
> ```js
> this.connectionManager.refreshTypeParser(dTypes)
> ```

I'm getting the same error: "Base constructors must all have the same return type.". Any hints?
